python/lsst/ts/observatory/control/auxtel/atcs.py

```diff
@@ -0,0 +1,1836 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+
+__all__ = ["ATCS", "ATCSUsages"]
+
+import enum
+import types
+import asyncio
+
+import numpy as np
+import astropy.units as u
+from astropy.coordinates import AltAz, ICRS, EarthLocation, Angle
+from astroquery.simbad import Simbad
+
+from ..remote_group import RemoteGroup, Usages
+from ..utils import subtract_angles, parallactic_angle
+
+from lsst.ts import salobj
+from lsst.ts.idl.enums import ATPtg, ATDome, ATPneumatics, ATMCS
+
+
+class RotType(enum.IntEnum):
+    """Defines the different types of rotator strategies.
+
+    Sky: Sky position angle strategy. The rotator is positioned with respect
+         to the North axis so rot_angle=0. means y-axis is aligned with North.
+         Angle grows clock-wise.
+
+    Parallactic: This strategy is required for taking optimum spectra with
+                 LATISS. If set to zero, the rotator is positioned so that the
+                 y-axis (dispersion axis) is aligned with the parallactic
+                 angle.
+
+    Physical: This strategy allows users to select the physical rotator angle
+              (in the reference frame of the telescope).
+    """
+
+    Sky = 0
+    Parallactic = 1
+    Physical = 2
+
+
+class ATCSUsages(Usages):
+    """ATCS usages definition.
+
+    Notes
+    -----
+
+    Additional usages definition:
+
+    Slew: Enable all slew operations.
+
+    StartUp: Enable startup operations.
+
+    Shutdown: Enable shutdown operations.
+
+    PrepareForFlatfield: Enable preparation for flat-field.
+    """
+
+    Slew = 1 << 3
+    StartUp = 1 << 4
+    Shutdown = 1 << 5
+    PrepareForFlatfield = 1 << 6
+
+    def __iter__(self):
+
+        return iter(
+            [
+                self.All,
+                self.StateTransition,
+                self.MonitorState,
+                self.MonitorHeartBeat,
+                self.Slew,
+                self.StartUp,
+                self.Shutdown,
+                self.PrepareForFlatfield,
+            ]
+        )
+
+
+class ATCS(RemoteGroup):
+    """High level library for the Auxiliary Telescope Control System
+
+    This is the high level interface for interacting with the CSCs that
+    control the Auxiliary Telescope. Essentially this will allow the user to
+    slew and track the telescope.
+
+    Parameters
+    ----------
+    domain: `salobj.Domain`
+        Domain to use of the Remotes. If `None`, create a new domain.
+    log : `logging.Logger`
+        Optional logging class to be used for logging operations. If `None`,
+        creates a new logger. Useful to use in salobj.BaseScript and allow
+        logging in the class use the script logging.
```

Comment:
Are you saying that if a subclass of BaseScript calls this, then it should pass its own log attribute? If so, I agree; that way log messages will be output using SAL.

What are the use cases for not passing an existing log?

Reply:
When you are not running this from a script, e.g. a jupyter notebook...